### ISSUE TITLE
Wrong class id in --save-txt, but correct class name in image annotation

### ISSUE THREAD

**Issue by KristofferK** [ultralytics/yolov5#6009]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Detection

### Bug

When I use the detect.py my images do get the correct annotation. E.g eggs get an egg annotations, larvae gets an larvae annotation, etc.

However, when using --save-txt, the class ids do not match the corresponding indices from the **names**-key in my **data_config.yaml**.

My data_config.yaml, has this:
**names: ['Eclosed egg', 'Egg', 'Larvae', 'Lump']**

Now I use detect.py, with a image that includes 2 eggs and 1 larvae.
The resulting image, contains 3 annotations. 2 eggs (class-id 1) annotations, 1 larvae (class-id 2) annotation. That is good.

Now when I go the generated labels directory and open the .txt file, it contains the following:
```
1 0.5215 0.696875 0.0203333 0.01525
1 0.397833 0.619 0.0216667 0.01
3 0.289333 0.4415 0.0213333 0.02
```

So in the image, it correctly annotated the larvae as an larvae (class id 2), but in the txt it writes class id 3 instead of class 2.

If I run detect.py on another file, where it contains Lumps (class id 3), then these are marked as (class id 2).

To sum it up:
The generated annotation is correct. The generated .txt uses the wrong class id.

Is the class-id in the .txt file not supposed to match the index in my yaml file?

### Environment

- YOLO: YOLOv5  v6.0-154-g14fc8ec torch 1.8.2 CUDA:0 (NVIDIA GeForce RTX 3060, 12288MiB)
- OS: Windows 10
- Python: 3.8.12

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by KristofferK** [comment]
I have tried to modify detect.py to include class-id in the annotation. So when I use detect.py, it shows that.
Class id 0: Lump
Class id 1: Egg
Class id 2: Eclosed egg
Class id 3: Larvae

But in Yaml they are:
Class id 0: Eclosed egg
Class id 1: Egg
Class id 2: Larvae
Class id 3: Lump

What could cause the mismatch between the name ids here and the names I have defined in my .yaml config?

**Comment by glenn-jocher** [comment]
@KristofferK 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. We are unable to reproduce any issue with detect.py: text files indicate correct class indices and annotated images display correct class labels (i.e. 5 is bus, 0 is person).

<img width="1679" alt="Screenshot 2021-12-16 at 19 51 20" src="https://user-images.githubusercontent.com/26833433/146431131-228294bc-469f-4fa7-bd8a-659d5621c95b.png">

We've created a few short guidelines below to help users provide what we need in order to get started investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by KristofferK** [comment]
I think it was a problem with the labels cache, where the indicies were wrong. This also solves my issue with https://github.com/ultralytics/yolov5/issues/5880 where the metrics in val.py did not match the ones from train.py. It should be fine now. Thanks.

**Comment by glenn-jocher** [comment]
@KristofferK if you can help us reproduce the issue we can start investigating, without a reproducible example there's nothing for us to act on.

**Comment by chrzc** [comment]
> 

Hi, I also meet this problem. I use the demo, COCO val
`!python val.py --weights yolov5s.pt --data coco.yaml --img 640 --iou 0.65 --half`
exp.  image_id  298251 
the picture label is true(zebra), but the category_id in predictions.json is 24 

## Environment
YOLO:YOLOv5 🚀 v6.0-193-gdb1f83b torch 1.10.1 CUDA:0 (NVIDIA GeForce RTX 3090, 24268MiB)
OS: Ubuntu 20.04
Python: 3.8.8
Torch: 1.10.1

**Comment by glenn-jocher** [comment]
@chrzc predictions.json has nothing to do with --save-txt and uses a different COCO indexing:
https://github.com/ultralytics/yolov5/blob/0cf932bf6346909189c53b375ef97551bb0c2326/val.py#L174
